### PR TITLE
Integrate SESOF MagRepack

### PR DESCRIPTION
**Pull request by kveberg**
Adds SESOF MagRepack to ACE Anvil.
- Please somehow note the contributions of Badger 2-3 to this mod -https://github.com/JBraunsmaJr
- This mod presently does not work with consoles because it hooks into drag-and-drop they do not have

### CODE REVIEW

**Comment by Kexanone** [comment]
I updated the folder structure, code style and added ACE prefixes. Make sure to add all contributors to [AUTHORS](https://github.com/acemod/ACE-Anvil/blob/master/AUTHORS). Also it would be good to have a short entry in the [docs](https://github.com/acemod/ACE-Anvil/tree/master/docs/src/components) for the mod with a hint that it originally was `SESOF MagRepack` and that it's not console compatible.

File: addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c [hunk 1]
```diff
@@ -0,0 +1,137 @@
+//------------------------------------------------------------------------------------------------
+modded class SCR_PlayerController
```
[the remark this answers is not in the document]
Comment (kveberg) [reply addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c:2]: I can add that, of course - it was my impression, however, that this is not the way: https://community.bistudio.com/wiki/Arma_Reforger:Scripting_Modding

[the remark this answers is not in the document]
Comment (Kexanone) [reply addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c:2]: I would argue it's more concise to have the base class there. Besides when you create a modded version via a copy of the original, you have it there anyway.

[the remark this answers is not in the document]
Comment (Kexanone) [reply addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c:2]: Besides, BI doesn't seem to be clear on which version to use either. Just have a look at the last section about debug process.

File: addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+			Print("Repacking exhausted fromMag.");
+			Print("Added " + (fromCount) + " rounds to toMag. It now holds " + toMag.GetAmmoCount() + ". The remainder in fromMag is " + fromMag.GetAmmoCount() + ".");
+						
+			// Reinsert toMag in optimal slot so things stack correctly
+			if (managerComp.TryRemoveItemFromInventory(toEntity, storageTo))
+			{
+				if (!managerComp.TryInsertItemInStorage(toEntity, managerComp.FindStorageForItem(toEntity, EStoragePurpose.PURPOSE_ANY)))
+				{
+					Print("Was unable to successfully add toMag back into storage. If it could not be dropped on the ground, it was deleted.", LogLevel.WARNING);
+					if (!managerComp.TryRemoveItemFromInventory(toEntity)) {SCR_EntityHelper.DeleteEntityAndChildren(toEntity);}
```
[the remark this answers is not in the document]
Comment (kveberg) [reply addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c:43]: This is a question of functionality, so it is up to you guys really. TryRemoveItemFromInventory will leave the mag on the ground, and so when you repack you could always pick it up again and fill it with new ammo. I intended to make a drop pouch to go with this but never got around to it (won't likely). Your suggestion there will of course just vanish the entire thing from existence the moment its empty. Which do you prefer for ACE?

File: addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		
+		if (!managerComp.IsValid())
+		{
+			Print("Invalid manager", LogLevel.ERROR);
+			return;
+		}
+		
+		if (!storageTo.IsValid())
+		{
+			Print("Invalid storageTo", LogLevel.ERROR);
```
[the remark this answers is not in the document]
Comment (kveberg) [reply addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c:121]: Apparently, yes. Fixed.

File: addons/magrepack/scripts/Game/ACE_MagRepack/UI/Inventory/SCR_InventoryMenuUI.c [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+		
+		fromInventoryItemComponent = m_pSelectedSlotUI.GetInventoryItemComponent();
+		fromEntity = fromInventoryItemComponent.GetOwner();
+		
+		// If we are dropping on the same slot we dragged from, only repack if the slot holds a stack
+		if (m_pSelectedSlotUI == m_pFocusedSlotUI)																				
+		{
+			if (m_pFocusedSlotUI.IsStacked())
+			{	
+				ref array<IEntity> items = {};
```
[the remark this answers is not in the document]
Comment (kveberg) [reply addons/magrepack/scripts/Game/ACE_MagRepack/UI/Inventory/SCR_InventoryMenuUI.c:58]: Right you are!

File: addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+		}
+		
+		if (!storageFrom.IsValid())
+		{
+			Print("Invalid storageFrom", LogLevel.ERROR);
```
[the remark this answers is not in the document]
Comment (kveberg) [reply addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c:129]: Yes, two actually. Fixed.

**Comment by kveberg** [comment]
I thought I had fixed all these, but apparently had not. Should be good now.

File: addons/magrepack/scripts/Game/ACE_MagRepack/UI/Inventory/SCR_InventoryMenuUI.c [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+		{
+			super.MoveItemToStorageSlot();
+		}
+	}
+
+	//------------------------------------------------------------------------------------------------
+	//! Checks to ensure we have two magazines that should be repacked, and passes the necessary bits to ACE_MagRepack_RpcAsk_Repack()-function in SCR_PlayerController
+	bool ACE_MagRepack_Repack()
+	{
+		Print("------- Attempting to repack magazines. -------", LogLevel.NORMAL);
```
Comment (Kexanone) [inline addons/magrepack/scripts/Game/ACE_MagRepack/UI/Inventory/SCR_InventoryMenuUI.c:17]:
```suggestion
		Print("------- Attempting to repack magazines. -------", LogLevel.DEBUG);
```

**Review by Kexanone — CHANGES_REQUESTED** [review]

File: addons/magrepack/scripts/Game/ACE_MagRepack/UI/Inventory/SCR_InventoryMenuUI.c [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+		BaseInventoryStorageComponent fromItemStorageComponent = m_pCallBack.m_pStorageFrom.GetCurrentNavigationStorage();
+		BaseInventoryStorageComponent toItemStorageComponent = m_pFocusedSlotUI.GetAsStorage();	
+		
+		SCR_PlayerController playerController = SCR_PlayerController.Cast(GetGame().GetPlayerController());
+		
+		RplComponent rpl = RplComponent.Cast(fromMag.GetOwner().FindComponent(RplComponent));
+		
+		if (rpl.IsMaster())
+		{
+			m_PlayerController.ACE_MagRepack_Repack(fromMag, toMag, m_InventoryManager, toItemStorageComponent, fromItemStorageComponent);
```
Comment (Kexanone) [inline addons/magrepack/scripts/Game/ACE_MagRepack/UI/Inventory/SCR_InventoryMenuUI.c:162]:
```suggestion
			playerController.ACE_MagRepack_Repack(fromMag, toMag, m_InventoryManager, toItemStorageComponent, fromItemStorageComponent);
```

**Review by Kexanone — CHANGES_REQUESTED** [review]

File: addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c [hunk 8]
```diff
@@ -0,0 +1,142 @@
+//------------------------------------------------------------------------------------------------
+modded class SCR_PlayerController : PlayerControllerClass
```
Comment (Kexanone) [inline addons/magrepack/scripts/Game/ACE_MagRepack/Player/SCR_PlayerController.c:2]:
```suggestion
modded class SCR_PlayerController : PlayerController
```

**Review by Kexanone — CHANGES_REQUESTED** [review]

**Review by Kexanone — APPROVED** [review]

**Pull request merged**